From c220115428e3de8f189afccadf6dc823825429fa Mon Sep 17 00:00:00 2001
From: wangjufeng <wangjufeng@huawei.com>
Date: Fri, 1 Nov 2019 16:48:49 +0800
Subject: [PATCH] fix syntax error

---
 Makefile.inc | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile.inc b/Makefile.inc
index e05f3a9..f4d895d 100644
--- a/Makefile.inc
+++ b/Makefile.inc
@@ -41,7 +41,7 @@ ifndef SYSTEMD
 	else
 		ifeq ($(shell systemctl --version >/dev/null 2>&1 && echo 1), 1)
 			SYSTEMD = $(shell systemctl --version 2> /dev/null | \
-				sed -n 's/systemd \([0-9]*\).*/\1/p')
+				sed -n 's/systemd\([0-9]*\).*/\1/p' | awk '{print $$1}')
 		endif
 	endif
 endif
-- 
1.8.3.1

