From 4e43314afe869f70f96d58dc978d81e104acb650 Mon Sep 17 00:00:00 2001
From: wangjufeng <wangjufeng@huawei.com>
Date: Fri, 1 Nov 2019 16:48:49 +0800
Subject: [PATCH 04/15] fix syntax error

---
 Makefile.inc | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile.inc b/Makefile.inc
index d0ec9b4..9944aa1 100644
--- a/Makefile.inc
+++ b/Makefile.inc
@@ -43,7 +43,7 @@ ifndef SYSTEMD
 	else
 		ifeq ($(shell systemctl --version >/dev/null 2>&1 && echo 1), 1)
 			SYSTEMD = $(shell systemctl --version 2> /dev/null | \
-				sed -n 's/systemd \([0-9]*\).*/\1/p')
+				sed -n 's/systemd\([0-9]*\).*/\1/p' | awk '{print $$1}')
 		endif
 	endif
 endif
-- 
1.8.3.1

