From 0008a2343765ab689cfc66acd8fc84630611d796 Mon Sep 17 00:00:00 2001
From: root <root@localhost.localdomain>
Date: Fri, 22 Mar 2019 19:37:43 +0800
Subject: [PATCH 1/5] change log level to info if alua is not support by
 storage  server

---
 libmultipath/prioritizers/alua.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/libmultipath/prioritizers/alua.c b/libmultipath/prioritizers/alua.c
index 1400b11..6b69982 100644
--- a/libmultipath/prioritizers/alua.c
+++ b/libmultipath/prioritizers/alua.c
@@ -127,7 +127,7 @@ int getprio (struct path * pp, char * args, unsigned int timeout)
 	} else {
 		switch(-rc) {
 		case ALUA_PRIO_NOT_SUPPORTED:
-			condlog(0, "%s: alua not supported", pp->dev);
+			condlog(3, "%s: alua not supported", pp->dev);
                         rc = 0;
 			break;
 		case ALUA_PRIO_RTPG_FAILED:
-- 
2.19.1

